<!--
################################################################################
# ProTN: an integrative pipeline for complete analysis of proteomics           # 
# data from mass spectrometry                                                  #
# Laboratory of RNA and Disease Data Science, University of Trento             #
# Developer: Gabriele Tomè                                                     #
# Issue at: https://github.com/gabrieletome/proTN_shiny/issues                         #
# PI: Dr. Toma Tebaldi, PhD                                                    #
################################################################################
-->
<img src="images/logo_scritta.svg" width="15%" style="display:block;margin-left:auto;margin-right:auto">
<h4>PhosProTN with proteome backgorund is an integrative pipeline for phosphoproteomic analysis of DDA experimental data
	obtained from MS. It perform a complete analysis of the raw files from Proteome Discoverer (PD) or MaxQuant (MQ),
	with their biological interpretation, enrichement and network analysis. PhosProTN analyse the phosphoproteomic data
	at peptide level using as background the proteome analysis of the same conditions.
</h4>
<div id="workflow-of-protn" class="section level2">
	<ul id="collapse_ul" class="collapsible">
		<li id="collapse_li" class="active">
			<div class="collapsible-header">
				<div style="width: 95%;">
					<h3>1. Workflow of PhosProTN with proteome background</h3>
					<hr id="linesubtitle">
				</div>
				<i id="plus_elem" class="fa fa-plus"></i>
			</div>
			<div class="collapsible-body">
				<p>The ProTN workflow is divided into preprocessing, differential analysis and biological
					interpretation. During
					preprocessing, input data are filtered, imputed, normalized, and optionally batch-corrected. Next,
					differential analysis
					is performed in parallel at peptide and protein resolution, based on users’ defined comparisons and
					considering compled
					designs. Finally, biological interpretation of differentially expressed proteins is performed,
					including functional
					enrichment analysis, and detection of communities in protein-protein interaction networks, and
					kinome tree perturbation
					analysis.
				</p>

				<img src="images/Workflow_PhosProTN.svg" width="100%" alt="">
				<div id="set-settings-for-the-execution-and-read-the-raw-data-from-loaded-files" class="section level3">
					<h3>01. Define analysis settings and load input data files
					</h3>
					<p>PhosProTN analyses the results of</p>
					<ul>
						<li>Proteome Discoverer</li>
						<li>MaxQuant</li>
					</ul>
					<p>Additional details on the input can be found in section <a
							href="#parameters-required-before-the-execution-phos-protn">2. Details on the input
							parameters and
							files</a></p>
				</div>
				<div id="normalization-of-the-intensities" class="section level3">
					<h3>02. Normalization and imputation of raw intensities</h3>
					<p>Intensities are log2 transformed and normalized with DEqMS <span class="citation">(Zhu
							2022)</span>. At the peptide level, the normalization is performed with the function
						equalMedianNormalization, which normalizes
						intensity distributions in samples so that they have median equal to 0.
					</p>
					<p>At the protein level, this operation is executed by the function medianSweeping, that applies the
						same median normalization
						used for peptides, but also summarizes peptide intensities into protein relative abundances by
						the median sweeping
						method.
					</p>
					<p>The principal method is based on the PhosR package <span class="citation">(Kim et
							al. 2021)</span> that performs a complex and well-balanced imputation of the data based on
						the association between replicates and
						conditions. As a backup method, ProTN uses a Gaussian round imputation, for condition with only
						1 replicate.
					</p>
					<p>In this step, many figure can be generated regarding information about pre-process, normalization
						and imputation.
						An example from the case study is the PCA based on the protein abundances below.
					</p>
					<img src="images/figures/PCA_readme_protn.png" width="55%"
						style="margin-left:auto;margin-right:auto;" />
				</div>
				<div id="statistical-differential-analysis" class="section level3">
					<h3>03. Differential analysis</h3>
					<p>Differential analysis is applied to both proteins and peptides, to identify significant
						differences. Two slightly
						different methodologies are applied: the DEqMS package <span class="citation">(Zhu
							2022)</span>,is used for proteins. DEqMS is developed on top of Limma, but the method
						estimates different prior
						variances for
						proteins quantified by different numbers of PSMs/peptides per protein, therefore achieving
						better accuracy.For
						single
						peptides, the Limma package is used
						<span class="citation">(Ritchie et al. 2015)</span>.
					</p>
					<ul>
						<li>
							<p><strong>Compile the comparison table</strong>: The table have 2 columns: </p>
							<ul>
								<li><strong>Formule</strong> column (<code>REQUIRED</code>): The formulas need to follow
									the syntax of <a href="https://bioinf.wehi.edu.au/limma/">Limma</a> (Ex:
									"<i>cancer-normal</i>").
								</li>
								<li><strong>Name</strong> column (<code>OPTIONAL</code>): personalized name assign to
									the comparison. (Ex: "<i>cancer_vs_normal</i>")</li>
							</ul>
						</li>
					</ul>
					<p>Limma and DEqMS calculate differentially expressed peptides and proteins (DEPs) for each
						comparison specified in the
						design file parameter. Each peptide or protein can be selected as differential based on
						different parameters: the log2
						Fold Change, the P.Value, the adjusted P.Value and the log2 expression. In ProTN, a
						protein/peptide is significant if
						passing thresholds on these parameters, set by the user. A protein/peptide for each comparison
						can be Up-regulated or
						Down-regulated. It is Up-regulated if:
					</p>
					<ul>
						<li>
							<p>the log2 FC is higher than the Fold Change threshold (FC &gt;
								Log2 FC thr),
							</p>
						</li>
						<li>
							<p>the Adj.P.Value or P.Value is lower than the threshold (P.Value &lt; P.Value
								thr),
							</p>
						</li>
					</ul>
					<p>It is Down-regulated if:</p>
					<ul>
						<li>
							<p>the log2 FC is lower of the Fold Change threshold (FC &lt; -Log2
								FC thr),
							</p>
						</li>
						<li>
							<p>the Adj.P.Value or P.Value is lower than the threshold (P.Value &lt; P.Value
								thr),
							</p>
						</li>
					</ul>
					<p>In the output, for each comparison, this distinction is reported in
						the “<em>class</em>” column, which assumes value “+” if is up-regulated,
						“-” if down-regulated and “=” if it is not significant.
					</p>
					<p>Various figures are generated, first a bar plot that graphically
						represents the DEPs identified. Followed by comparison-specific volcano
						plots.
					</p>
					<img src="images/figures/DEPs_readme_protn.png" width="49%"
						style="margin-left:auto;margin-right:auto" />
					<img src="images/figures/protein_vulcano.png" width="49%"
						style="margin-left:auto;margin-right:auto" />
				</div>
				<div id="report-creation-and-download-of-the-results" class="section level3">
					<h3>04. Report creation and download of the results</h3>
					<p>The results are summarized in a web-page HTML report. Other than
						this, the experiment is described by a large number of files, a
						description of each file generated can be found in section <a
							href="#details-on-the-output-files-phos">3. Details on the output files</a>.
						All the files are group in a <em>zip</em> file and downloaded.
					</p>
					<img src="images/figures/protn_report.html(screen).png" width="75%"
						style="display:block;margin-left:auto;margin-right:auto" />

				</div>
				<h3>ADDITIONAL STEPS:</h3>
				<div id="batch-effect-correction" class="section level3">
					<h3>B1. Batch Effect correction</h3>
					<p>If required by the experiment, a batch correction step can be applyed
						using <a href="https://www.bioconductor.org/packages/release/bioc/html/proBatch.html"
							style="font-size: 11pt;">proBatch</a> <span class="citation">(Cuklina et
							al. 2018)</span>. The batches need to be defined in the
						<strong>sample annotation</strong> file where an additional column describe the batches.
						required.
					</p>
				</div>
				<div id="enrichment-analysis-of-the-differentially-expressed-proteins" class="section level3">
					<h3>E1. Enrichment analysis of the Differentially Expressed
						Proteins
					</h3>
					<div style="display: flex;">
						<p>The biological interpretation of the Differentially Expressed
							Proteins starts with the enrichment step. To execute this analysis,
							ProTN uses <a href="https://maayanlab.cloud/Enrichr/">EnrichR</a> <span
								class="citation">(Jawaid 2022)</span>, a popular tool that searches on a large number of
							data sets to obtain information about many functional categories.
							EnrichR organises its hundreds of data sources in 8 sections: Transcription, Pathways,
							Ontologies, Diseases/Drugs, Cell
							Types, Misc, Legacy, and Crowd.
						</p>

						<p><a href="https://maayanlab.cloud/Enrichr/"><img src="images/enrichr_logo.png" width="200px"
									alt="" /></a></p>
					</div>
					<p>Each comparison defined in the differential analysis stage can result in 3 sets of proteins: the
						Up-regulated (called
						Up), the Down-regulated (called Down), and the merge of the two (called all). ProTN provides for
						each term statistical
						parameters like P.Value, fdr, odds ratio, overlap size.
					</p>
					<p>ProTN creates an RData of the complete enrichment data frame, allowing the user an easy import in
						R to perform further
						analysis. ProTN also generates an Excel file, containing only the significantly enriched terms,
						as defined by user
						settings.
					</p>
					<p>A term to be significative need to have:</p>
					<ul>
						<li>
							<p>a Fdr or P.Value lower of P.Value thr for enrichment (P.Value &lt;
								P.Value thr for enrichment),
							</p>
						</li>
						<li>
							<p>an Overlap Size higher than Overlap size thr for enrichment
								(Overlap Size &gt; Overlap size thr for enrichment).
							</p>
						</li>
					</ul>
					<p>ProTN displays top significant enrichments based on specific annotation datasets or keywords
						selected by the user.
					</p>
					<img src="images/figures/enrichment_readme_protn.png" width="50%"
						style="display:block;margin-left:auto;margin-right:auto" />
				</div>
				<div id="kinase-tree-analysis" class="section level3">
					<h3>K1. Activity kinase tree analysis of the Differentially Expressed Phosphosite
					</h3>
					<div style="display: flex;">
						<p>In phospho-proteomic it extremely useful
							to study the activation status of the kinase based on the differentially expressed
							substrate idenfied by the differential analysis.
							For each comparison, PhosProTN predicts the activation state of the kinases using PhosR
							<span class="citation">(Kim et al. 2021)</span>. PhosR provides a kinase-substrate
							relationship
							score, and on that it prioritises potential kinases that could be responsible for the
							phosphorylation
							change of phosphosite on the basis of kinase recognition motif and phosphoproteomic
							dynamics.
						</p>
						<p><a href="http://phanstiel-lab.med.unc.edu/CORAL/"><img src="images/coral-logo-black.png"
									width="200px" alt="" /></a></p>
					</div>
					<p>The activity score provide by PhosR is used to generated a graphical versione of the human kinome
						tree using
						<a href="http://phanstiel-lab.med.unc.edu/CORAL/">CORAL</a> <span class="citation">(Metz K.S. et
							al. 2018)</span>,
						a web shiny app for visualizing both quantitative and qualitative data. It generates
						high-resolution scalable vector graphic files suitable for publication
						without the need for refinement in graphic editing software.
					</p>
					<img src="images/figures/TA_kinase_Tree_CORAL-cropped.svg" width="50%"
						style="display:block;margin-left:auto;margin-right:auto" />
				</div>
				<div id="protein-protein-interaction-network-analysis-of-differentially-expressed-proteins"
					class="section level3">
					<h3>N1. Protein-Protein Interaction network analysis of Differentially
						Expressed Phosphosite
					</h3>
					<div style="display: flex;">
						<p>ProTN performs Protein-Protein Interaction (PPI) network analysis on differentially expressed
							proteins. PPIs are
							essential in almost all processes of the cell, and crucial for understanding cell physiology
							in different states.
							For each comparison, ProTN analyses the interaction between the DEPs
							using <a href="https://string-db.org/">STRING</a> <span class="citation">(Szklarczyk et al.
								2021)</span>.
						</p>
						<p><a href="https://string-db.org/"><img src="images/string_logo_2015.png" width="200"
									alt=""></a></p>
					</div>
					<p>The species-specific database is retrieved from STRING server, an
						accurate analysis discover all the interactions and an iGraph <span class="citation">(Csardi and
							Nepusz 2006)</span> network is generated.
						Later, the proteins are clustered via iGraph function which identify
						dense subgraph by optimizing modularity score.
					</p>
					<p>Since the network can vary a lot on composition, two ggplot layout
						are used: Fruchterman-Reingold algorithm and the Kamada-Kawai
						algorithm.
					</p>
					<img src="images/figures/stringdb_readme_protn.png" width="75%"
						style="display:block;margin-left:auto;margin-right:auto" />
				</div>
			</div>
		</li>
	</ul>
</div>
<div id="parameters-required-before-the-execution-phos-protn" class="section level2">
	<ul id="collapse_ul" class="collapsible">
		<li id="collapse_li">
			<div class="collapsible-header">
				<div style="width: 95%;">
					<h3>2. Details on the input parameters and files</h3>
					<hr id="linesubtitle">
				</div>
				<i id="plus_elem" class="fa fa-plus"></i>
			</div>
			<div class="collapsible-body">
				<!--TODO: ADD DIA analysis-->
				<ul>
					<li>
						<p><code>Title of Analysis</code>: title of the experiment. It will
							be the title of the web page report.
						</p>
					</li>
					<li>
						<p><code>Brief Description</code>: description of the current
							experiment. It is the first paragraph of the report.
						</p>
					</li>
					<li><code>Software Analyzer</code>: determine which software was used to identify peptides and
						proteins.
						<ul>
							<li>Proteome Discoverer</li>
							<li>MaxQuant using evidence.txt</li>
							<li>MaxQuant using peptides.txt and proteinGroups.txt</li>
							<li>Spectronaut</li>
							<li>FragPipe</li>
						</ul>
					</li>
					<li>
						<h4>File required for Proteome Discoverer</h4>

						<ul>
							<li><code>Annotation file</code>: This file provides metadata for the samples analyzed. It
								must be an Excel file with the following required columns:
								<table>
									<thead>
										<tr>
											<th>Column Name</th>
											<th>Description</th>
										</tr>
									</thead>
									<tbody>
										<tr>
											<td><code>File ID</code></td>
											<td>Identifier used in column headers of the peptide file.</td>
										</tr>
										<tr>
											<td><code>Condition</code></td>
											<td>Experimental group name. Used for comparisons.</td>
										</tr>
										<tr>
											<td><code>Sample</code></td>
											<td>Clean sample name used downstream.</td>
										</tr>
										<tr>
											<td><code>color</code></td>
											<td>(Optional) Plot color. Defaults are applied if missing.</td>
										</tr>
										<tr>
											<td><code>batch</code></td>
											<td>(Optional) Batch ID for batch effect correction.</td>
										</tr>
									</tbody>
								</table>
							</li>
							<li><code>Peptides file</code>: Excel table with annotated peptides and abundance values.
								<table>
									<thead>
										<tr>
											<th>Column Name</th>
											<th>Description</th>
										</tr>
									</thead>
									<tbody>
										<tr>
											<td><code>Master Protein Accessions</code></td>
											<td>Maps peptide to protein; only first ID is kept.</td>
										</tr>
										<tr>
											<td><code>Annotated Sequence</code></td>
											<td>Amino acid sequence including PTM annotations.</td>
										</tr>
										<tr>
											<td><code>Modifications</code></td>
											<td>Post-translational modifications.</td>
										</tr>
										<tr>
											<td><code>Positions in Master Proteins</code></td>
											<td>Position of peptide in the protein sequence.</td>
										</tr>
										<tr>
											<td><code>Abundance: &lt;File ID&gt;</code></td>
											<td>Intensity/abundance for each sample. One column per sample.</td>
										</tr>
									</tbody>
								</table>
							</li>
							<li><code>Proteins file</code>: Excel table containing descriptive and accession information
								for proteins.
								<table>
									<thead>
										<tr>
											<th>Column Name</th>
											<th>Description</th>
										</tr>
									</thead>
									<tbody>
										<tr>
											<td><code>Accession</code></td>
											<td>Unique protein identifier, used to join with peptide file.</td>
										</tr>
										<tr>
											<td><code>Description</code></td>
											<td>Descriptive string, e.g., from UniProt.</td>
										</tr>
									</tbody>
								</table>
							</li>
							<li><code>PSM file</code>: only for Phospho dataset. For Proteome background is not
								required.
								<table>
									<thead>
										<tr>
											<th>Column Name</th>
											<th>Description</th>
										</tr>
									</thead>
									<tbody>
										<tr>
											<td><code>ptmRS: Best Site Probabilities</code></td>
											<td>Used to resolve phosphosite ambiguity.</td>
										</tr>
										<tr>
											<td><code>Precursor Abundance</code></td>
											<td>Abundance value used to filter invalid entries.</td>
										</tr>
										<tr>
											<td><code>Master Protein Accessions</code></td>
											<td>Matches protein IDs for mapping.</td>
										</tr>
										<tr>
											<td><code>Annotated Sequence</code></td>
											<td>Used to resolve conflicting PTM assignments.</td>
										</tr>
									</tbody>
								</table>
							</li>
						</ul>
					</li>
					<li>
						<h4>File required for MaxQuant</h4>

						<ul>
							<li><code>Annotation file</code>: This file provides metadata for the samples analyzed. It
								must be an Excel file with the following required columns:
								<table>
									<thead>
										<tr>
											<th>Column Name</th>
											<th>Description</th>
										</tr>
									</thead>
									<tbody>
										<tr>
											<td><code>Condition</code></td>
											<td>Experimental condition (e.g. Control, Treated). Used for group
												comparison.</td>
										</tr>
										<tr>
											<td><code>Sample</code></td>
											<td>Sample identifier. Must match sample names in the peptide file.</td>
										</tr>
										<tr>
											<td><code>color</code></td>
											<td>(Optional) Color associated with the condition. If not present, default
												colors are assigned.</td>
										</tr>
										<tr>
											<td><code>batch</code></td>
											<td>(Optional) Batch ID for batch effect correction. Required if batch
												correction is enabled.</td>
										</tr>
									</tbody>
								</table>
							</li>

							<li><strong>Evidence pipeline</strong>:
								<ul>
									<li><code>evidence.txt</code>: This is a TSV/CSV file containing peptide-level
										quantification data. <strong>Required columns:</strong>
										<table>
											<thead>
												<tr>
													<th>Column Name</th>
													<th>Description</th>
												</tr>
											</thead>
											<tbody>
												<tr>
													<td><code>Sequence</code></td>
													<td>Amino acid sequence of the peptide.</td>
												</tr>
												<tr>
													<td><code>Modifications</code></td>
													<td>PTMs of the peptide.</td>
												</tr>
												<tr>
													<td><code>Gene names</code></td>
													<td>Gene symbol associated with the peptide.</td>
												</tr>
												<tr>
													<td><code>Protein names</code></td>
													<td>Protein description. If missing, will be merged from annotation
														file.</td>
												</tr>
												<tr>
													<td><code>Leading razor protein</code></td>
													<td>UniProt accession. Used for annotation enrichment.</td>
												</tr>
												<tr>
													<td><code>Raw file</code></td>
													<td>File/sample ID. Must match entries in the annotation file.</td>
												</tr>
												<tr>
													<td><code>Intensity</code></td>
													<td>Peptide intensity value. Used for quantification.</td>
												</tr>
												<tr>
													<td><code>Leading proteins</code></td>
													<td>Used for filtering out contaminants (e.g. "CON_").</td>
												</tr>
											</tbody>
										</table>
									</li>
								</ul>
							</li>
						</ul>
					</li>
				</ul>
			</div>
		</li>
	</ul>
</div>
<div id="example-case-study---proteomics-of-mcf7-cells" class="section level2">
	<ul id="collapse_ul" class="collapsible">
		<li id="collapse_li">
			<div class="collapsible-header">
				<div style="width: 95%;">
					<h3>3. Example case study - Phosphoproteomics from Steger et al. (2016)</h3>
					<hr id="linesubtitle">
				</div>
				<i id="plus_elem" class="fa fa-plus"></i>
			</div>
			<div class="collapsible-body">
				<p><code>Phosphoproteomics reveals that Parkinson's disease kinase LRRK2 regulates a subset of Rab GTPases</code>
				</p>
				<p>PRIDE: <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD003071">PXD003071</a>.
				<p>Des: Steger M, Tonelli F, Ito G, Davies P, Trost M, Vetter M, Wachter S, Lorentzen E, Duddy G, Wilson
					S, Baptista MA, Fiske BK, Fell MJ, Morrow JA, Reith AD, Alessi DR, Mann M. Phosphoproteomics reveals
					that Parkinson's disease kinase LRRK2 regulates a subset of Rab GTPases. Elife. 2016 Jan 29;5.</p>
				<p>Possible to download the case study in the Run tab.</p>
			</div>
		</li>
	</ul>
</div>

<div id="details-on-the-output-files-phos" class="section level2">
	<ul id="collapse_ul" class="collapsible">
		<li id="collapse_li">
			<div class="collapsible-header">
				<div style="width: 95%;">
					<h3>4. Details on the output files</h3>
					<hr id="linesubtitle">
				</div>
				<i id="plus_elem" class="fa fa-plus"></i>
			</div>
			<div class="collapsible-body">
				<ul>
					<li><code>report.html</code>: complete report of the analysis with all pics and results of the
						enrichment.</li>
					<li><code>db_results_proTN.RData</code>: RData object containing all the data produced during the
						execution ready for additional analyses in R.</li>
					<li><code>log_filter_read_function.txt</code>: Results filter applied during preprocessing step.
					</li>
					<li><strong>input_protn folder</strong>
						<ul>
							<li><code>Input files provided</code></li>
						</ul>
					</li>
					<li><strong>rdata folder</strong>
						<ul>
							<li><code>enrichment.RData</code>: RData object containing enrichment
								results based on differentially expressed proteins</li>
						</ul>
					</li>
					<li><strong>Tables folder</strong>
						<ul>
							<li><code>normalised_abundances.xlsx</code>: excel file containing
								abundance values generated by proTN. abundances are log2 transformed,
								normalized, imputed (and batch corrected). The file is organized in the
								following sheets:
								<ul>
									<li><strong>protein_per_sample</strong>: protein abundances per
										sample.</li>
									<li><strong>peptide_per_sample</strong>: peptide abundances per
										sample.</li>
									<li><strong>protein_per_condition</strong>: protein abundances per
										condition (average &amp; standard deviation), as defined in the Sample
										Annotation.</li>
									<li><strong>peptide_per_condition</strong>: peptide abundances per
										condition (average &amp; standard deviation), as defined in the Sample
										Annotation.</li>
								</ul>
							</li>
							<li><code>differential_expression.xlsx</code>: excel file containing the
								results of differential analysis, according to the contrasts defined in
								the Design file. The file is organized in the following sheets:
								<ul>
									<li><strong>protein_DE</strong>: protein differential expression results
										protein abundances per sample.</li>
									<li><strong>peptide_DE</strong>: peptide differential expression
										results. Annotation columns:</li>
									<li><strong>Accession</strong>: protein UniprotID</li>
									<li><strong>Description</strong>: protein description</li>
									<li><strong>GeneName</strong>: Gene Symbol</li>
									<li><strong>Peptide_Sequence</strong>: peptide sequence</li>
									<li><strong>Peptide_Modifications</strong>: peptide modifications</li>
									<li><strong>Peptide_Position</strong>: start and end position of the
										peptide within the protein sequence, defined UniprotID</li>
									<li><strong>Peptide_Tryptic</strong>: peptide tryptic digestion status
										(fully tryptic, N-semi tryptic, C-semi tryptic, non tryptic)</li>
									<li>Columns for each contrast:</li>
									<li><strong>class</strong>: defined according to the fold change,
										p-value and abundance thresholds specified in the input</li>
									<li><em>+</em> up-regulated protein/peptide</li>
									<li><em>-</em> down-regulated protein/peptide</li>
									<li><em>=</em> invariant protein/peptide</li>
									<li><strong>log2_FC</strong>: protein/peptide log2 transformed fold
										change</li>
									<li><strong>p_val</strong>: protein/peptide contrast p-value</li>
									<li><strong>p_adj</strong>: protein/peptide adjusted p-value (FDR after
										BH correction)</li>
									<li><strong>log2_expr</strong>: protein/peptide log2 average
										abundance</li>
								</ul>
							</li>
							<li><code>enrichment.xlsx</code>: excel file containing a selection
								of enrichment results starting from differentially expressed proteins.
								Terms are selected according to significance thresholds specified in the
								input (Default: adj.P.Value &lt; 0.05, Overlap Size &gt;= 5)</li>
						</ul>
					</li>
					<li><strong>Figures folder</strong>
					<li><strong>PDF version of all figure selected</strong></li>
					<li><strong>enrichment_plot.pdf</strong>: dot plot of top enriched
						terms based on differentially expressed proteins, divided in up- and
						down-regulated. Terms are filtered for keywords defined in the advanced
						options.</li>
		</li>
		<li><code>protein_vulcano</code> directory: Directory with all the vulcano plots based on the differential
			proteins.
		</li>
		<li><code>peptide_vulcano</code> directory: Directory with all the vulcano plots based on the differential
			peptide.
		</li>
		<li><code>STRINGdb</code> directory: Directory with figures from network
			analysis of differentially expressed proteins, based on STRINGdb
			protein-protein interactions. For each contrast, two files are
			generated:
			<ul>
				<li><strong>*_connection.txt</strong>: txt file with all edge of the network.</li>
				<li><strong>*_network.pdf</strong>: PDF version of STRINGdb network.</li>
			</ul>
		</li>
		<li>
			<code>KinaseTree</code> directory: figures and files with the kinase activity trees generated
			with PhosR and CORAL. For each deasign there are two files:
			<ul>
				<li>
					<code>TXT file</code>: text table with the identified actifity of each kinase.
				</li>
				<li>
					<code>SVG file</code>: vectorail image of the kinase tree generated with CORAL.
				</li>
			</ul>
		</li>
	</ul>
</div>
</li>
</ul>
</div>
<hr id="linesubtitle">
<div id="refs" class="references csl-bib-body hanging-indent">
	<div id="ref-igraph" class="csl-entry">
		Csardi, Gabor, and Tamas Nepusz. 2006. <span>“The Igraph Software
			Package for Complex Network Research”</span> Complex Systems: 1695. <a
			href="https://igraph.org">https://igraph.org</a>.
	</div>
	<br>
	<div id="ref-proBatch" class="csl-entry">
		Cuklina, Jelena, Chloe H. Lee, Evan G. Willams, Ben Collins, Tatjana
		Sajic, Patrick Pedrioli, Maria Rodriguez-Martinez, and Ruedi Aebersold.
		2018. <span>“Computational Challenges in Biomarker Discovery from
			High-Throughput Proteomic Data.”</span> <a
			href="https://doi.org/10.3929/ethz-b-000307772">https://doi.org/10.3929/ethz-b-000307772</a>.
	</div>
	<br>
	<div id="ref-enrichR" class="csl-entry">
		Jawaid, Wajid. 2022. <span>“enrichR: Provides an r Interface to
			’Enrichr’.”</span> <a
			href="https://CRAN.R-project.org/package=enrichR">https://CRAN.R-project.org/package=enrichR</a>.
	</div>
	<br>
	<div id="ref-PhosR" class="csl-entry">
		Kim, Hani Jieun, Taiyun Kim, Nolan J Hoffman, Di Xiao, David E James,
		Sean J Humphrey, and Pengyi Yang. 2021. <span>“PhosR Enables Processing
			and Functional Analysis of Phosphoproteomic Data”</span> 34. <a
			href="https://doi.org/10.1016/j.celrep.2021.108771">https://doi.org/10.1016/j.celrep.2021.108771</a>.
	</div>
	<br>
	<div id="ref-limma" class="csl-entry">
		Ritchie, Matthew E, Belinda Phipson, Di Wu, Yifang Hu, Charity W Law,
		Wei Shi, and Gordon K Smyth. 2015.
		<span>“<span></span>Limma<span></span> Powers Differential Expression
			Analyses for <span></span>RNA<span></span>-Sequencing and Microarray
			Studies”</span> 43: e47. <a
			href="https://doi.org/10.1093/nar/gkv007">https://doi.org/10.1093/nar/gkv007</a>.
	</div>
	<br>
	<div id="ref-coral" class="csl-entry">
		Kathleen S. Metz, Erika M. Deoudes, Matthew E. Berginski, Ivan Jimenez-Ruiz,
		Bulent Arman Aksoy, Jeff Hammerbacher, Shawn M. Gomez and Douglas H. Phanstie. 2018.
		<span>Coral: Clear and Customizable Visualization of Human Kinome Data</span>
		<span>Cell Systems</span>Volume 7, Issue 3, 347-350
		<a href="https://doi.org/10.1016/j.cels.2018.07.001">https://doi.org/10.1016/j.cels.2018.07.001</a>.
	</div>
	<br>
	<div id="ref-STRINGdb" class="csl-entry">
		Szklarczyk, Damian, Annika L Gable, Katerina C Nastou, David Lyon,
		Rebecca Kirsch, Sampo Pyysalo, Nadezhda T Doncheva, et al. 2021.
		<span>“The STRING Database in 2021: Customizable Protein-Protein
			Networks, and Functional Characterization of User-Uploaded
			Gene/Measurement Sets.”</span> 49.
	</div>
	<br>
	<div id="ref-DEqMS" class="csl-entry">
		Zhu, Yafeng. 2022. <span>“DEqMS: A Tool to Perform Statistical Analysis
			of Differential Protein Expression for Quantitative Proteomics
			Data.”</span>
	</div>
</div>